dipy/io/statefull_tractogram.py

```diff
@@ -0,0 +1,879 @@
+
+    def convert_data_per_point_for_trk(self):
+        """
+        Trk requires a specific way of writting metadata, attempt (heuristic)
+        at converting typical dictionnary to trk-friendly format
+        (WARNING)
+        """
+        for key in self._data_per_point:
+            for ind in range(len(self._data_per_point[key])):
+                for i, value in enumerate(self._data_per_point[key][ind]):
+                    if isinstance(value, numbers.Number):
+                        value = float(value)
+                    if not isinstance(value, (list, np.ndarray)):
+                        self._data_per_point[key][ind][i] = [value]
+
+    def get_space_attribute(self):
+        """ Getter for spatial attribute """
+        return self._affine, self._dimensions, \
+            self._voxel_sizes, self._voxel_order
+
+    def get_current_space(self):
+        """ Getter for the current space """
+        return self._space
+
+    def get_current_shift(self):
+        """ Getter for shift """
+        return self._shifted_origin
+
+    def get_streamlines(self):
+        """ Partially safe getter for streamlines """
+        return self._streamlines
+
+    def get_streamlines_copy(self):
+        """ Safe getter for streamlines (for slicing) """
+        return deepcopy(list(self._streamlines))
+
+    def set_streamlines(self, streamlines):
+        """ Modify streamlines. Creating a new object would be less risky.
+
+        Parameters
+        ----------
+        streamlines : list or ArraySequence (list and deepcopy recommanded)
+            Streamlines of the tractogram
+
+        Returns
+        -------
+        output : bool
+            Was the operation successful (did everything match)
+        """
+        if not isinstance(streamlines, (list, ArraySequence)):
+            raise TypeError('Streamlines MUST be a list or an ArraySequence')
+
+        if not _is_data_per_point_valid(streamlines, self._data_per_point) or \
+            not _is_data_per_streamline_valid(streamlines,
+                                              self._data_per_streamline):
+            raise ValueError('Invalid data_per_point or data_per_streamline')
+
+        logging.warning('Streamlines were modified, data still match')
+        self._streamlines = ArraySequence(streamlines)
+
+    def set_streamlines_and_data(self, streamlines,
+                                 overwrite_data=False,
+                                 data_per_point=None,
+                                 data_per_streamline=None):
+        """ Modify streamlines AND data at the same time. Creating a new object
+        would be less risky.
+
+        Setting one of both or both data at the same time is possible.
+        Using overwrite_data without data will erase data (with {})
+
+        Subsampling streamlines from ArraySequence should ALWAYS be done with a
+        deepcopy and pass as a list. ArraySequence views are broken !
```

Comment:
What do you mean by "ArraySequence views are broken!" Is it something that should be fixed in NiBabel?

Reply:
For the array sequence view. If you slices an ArraySequence (something like sub_streamlines = streamlines[0:10]) sub_streamlines.data contains ALL data from streamlines. I called it broken, maybe that's not the right word, it is more like "surprising behaviour" for a lot of people.

I think that this class is needed for Dipy 1.0 in July. Do you think this new class would be acceptable in its current state for Nibabel ?

I am not trying to push anything too fast, but when (if) this is merged there is a lot of work (that I am willing to do) to fix the current handling of spatial transformation and IO.
![3D_screenshot](https://user-images.githubusercontent.com/10820351/57028173-60be4380-6c0c-11e9-9021-29f4584cfef2.png)
All of these tractograms were computed from the example (and workflows) and they should be aligned (out of 13, 3 display at the right place in MI-Brain, only 2 out of these 3 have a adequate header)

Reply:
The main thing that is "limiting" right now is the support for lazy-loading and lazy-saving. From a "design" POV, it will be hard to maintain a state (like with the current PR) while also supporting lazy operations.

For example, say that we are lazy loading some streamlines. Then, for some reason, in the middle of iterating on the streamlines, some operation is done, like a call to `to_vox`. Then, what should be the behavior? Invalidate the iterator, I guess. This would make sense since we probably don't expect people to call a state changing function during iteration.

I think it would be doable, but would take longer. In the meantime, as @frheault just showed, most of the tutorials / functions are not consistent in Dipy. I would suggest the following plan, if you agree:

1- Integrate everything that is relevant to Nibabel based on the current design (after reviews and refactors if needed)
2- Fix tutorials, functions and examples in Dipy to be consistent and clear on what they expect.
3- [optional] Refactor StatefullTractogram to support lazy operations, if we can reliably handle it.

What do you think, @MarcCote ?

Reply:
That's a sobering picture!

Regarding the brokeness in nibabel, this is what I was mentioning here? https://github.com/nipy/nibabel/issues/729

I still owe nibabel a tutorial :-(

Reply:
@frheault I'm all for making it easier for the users. Those are great wrappers :+1: .

@jchoude I agree with that plan. I think it also goes well with what @arokem proposed below (i.e. test everything out in DIPY 1.0 then move what needs to be moved into Nibabel).

@arokem yes you are right about the issue you mentioned. I wasn't expecting people to call `.data` directly (I don't remember why we didn't make it private in the first place, though) but I guess it is useful after all. That should be addressed in Nibabel, though.